File: services/graphql-ims/src/models/returns.rs
use std::sync::Arc;

use async_graphql::{ComplexObject, Context, dataloader::Loader};
use chrono::{DateTime, Utc};
use graphql_core::PostgresDataLoader;
use graphql_crm::models::companies;
use uuid::Uuid;

use crate::models::sales_orders;

use super::enums::ReturnStatusEnum;

#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq)]
pub struct PrimaryKey(pub Uuid);

#[derive(Clone, Debug, PartialEq, Eq, async_graphql::SimpleObject, sqlx::FromRow)]
#[graphql(name = "ImsReturns", complex)]
pub struct Model {
    pub id: Uuid,
    pub return_number: String,
    #[graphql(skip)]
    pub sales_order_id: Option<Uuid>,
    #[graphql(skip)]
    pub client_id: Uuid,
    pub status: Option<ReturnStatusEnum>,
    pub reason: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[ComplexObject]
impl Model {
    async fn sales_order(
        &self,
        ctx: &Context<'_>,
    ) -> async_graphql::Result<Option<sales_orders::Model>> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        if let Some(id) = self.sales_order_id {
            Ok(loader.load_one(sales_orders::PrimaryKey(id)).await?)
        } else {
            Ok(None)
        }
    }
    async fn client(&self, ctx: &Context<'_>) -> async_graphql::Result<companies::Model> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        Ok(loader
            .load_one(companies::PrimaryKey(self.client_id))
            .await?
            .ok_or(async_graphql::Error::new("Unable to get product"))?)
    }
}

impl Loader<PrimaryKey> for PostgresDataLoader {
    type Error = Arc<sqlx::Error>;
    type Value = Model;

    async fn load(
        &self,
        keys: &[PrimaryKey],
    ) -> Result<std::collections::HashMap<PrimaryKey, Self::Value>, Self::Error> {
        let keys = keys.iter().map(|k| k.0).collect::<Vec<_>>();

        let results =
            sqlx::query_as::<_, Self::Value>("select * from ims.returns where id = ANY($1)")
                .bind(&keys)
                .fetch_all(&self.pool)
                .await?
                .into_iter()
                .map(|model| (PrimaryKey(model.id), model))
                .collect::<_>();

        Ok(results)
    }
}

File: src/graphql/queries/wms/returns.ts
import { graphql } from "@/lib/graphql/client";

/**
 * Query to fetch a single return item by its ID.
 */
export const getReturnItem = graphql(`
  query GetReturnItem($id: UUID!) {
    wms {
      returnItem(id: $id) {
        id
        returnNumber
        status
        reason
        createdAt
        updatedAt
        salesOrder {
          id
          orderNumber
        }
        client {
          id
          name
        }
      }
    }
  }
`);

/**
 * Query to fetch a list of returns with pagination.
 */
export const getReturns = graphql(`
  query GetReturns($limit: Int!, $page: Int!) {
    wms {
      returns(limit: $limit, page: $page) {
        id
        returnNumber
        status
        reason
        createdAt
        updatedAt
        salesOrder {
          id
          orderNumber
        }
        client {
          id
          name
        }
      }
    }
  }
`);
